src/Pate/Verification/PairGraph.hs

```diff
@@ -100,60 +66,166 @@ import qualified Pate.Verification.SymbolicExecution as PVSy
+--   overapproximation of all the differences that may exist between
+--   runs of the two programs.
+--
+--   As we compute the fixpoint we note cases where we discover a
+--   location where we cannot keep the control-flow syncrhronized, or
+--   where we discover some observable difference in the program
+--   behavior. These "program desyncrhonization" and "observable
+--   difference" occurences are ultimately the information we want to
+--   deliver to the user.
+--
+--   We additionally track situations where we cut off the fixpoint
+--   computation early as reportable situations that represent
+--   limitations of the analysis; such situations represent potential
+--   sources of unsoundness that may cause us to miss desyncronization
+--   or observable difference events.
 data PairGraph sym arch =
   PairGraph
-  { pairGraphDomains :: !(Map (PPa.BlockPair arch) (AbstractDomain sym arch))
-  , pairGraphGas :: !(Map (PPa.BlockPair arch, PPa.BlockPair arch) Gas)
-  , pairGraphWorklist :: !(Set (PPa.BlockPair arch)) -- TODO? priority queue of some kind?
+  { -- | The main data structure for the pair graph, which records the current abstract
+    --   domain value for each reachable graph node.
+    pairGraphDomains :: !(Map (GraphNode arch) (AbstractDomain sym arch))
+
+    -- | This data structure records the amount of remaining "gas" corresponding to each
+    --   edge of the program graph. Initially, this map is empty. When we traverse an
+    --   edge for the first time, we record it's initial amount of gas, which is later
+    --   decremented each time we perform an update along this edge in the future.
+    --   We stop propagating updates along this edge when the amount of gas reaches zero.
+  , pairGraphGas :: !(Map (GraphNode arch, GraphNode arch) Gas)
+
+    -- | The "worklist" which records the set of nodes that must be revisited.
+    --   Whenever we propagate a new abstract domain to a node, that node must
+    --   be revisited, and here we record all such nodes that must be examinied.
+    --
+    --   For now, this is just a set and we examine nodes in whatever order is defined
+    --   by their Ord instance. It may make sense at some point to use a more sophistocated
+    --   mechanism to determine the order in which to visit nodes.
+  , pairGraphWorklist :: !(Set (GraphNode arch))
+
+    -- | The set of blocks where this function may return to. Whenever we see a function
+    --   call to the given FunPair, we record the program point pair where the function
+    --   returns to here. This is used to tell us where we need to propagate abstract domain
+    --   information when visiting a ReturnNode.
+  , pairGraphReturnVectors :: !(Map (PPa.FunPair arch) (Set (PPa.BlockPair arch)))
+
+    -- TODO, I'm not entirely sure I love this idea of tracking error conditions in this
+    -- data structure like this.  It works for now, but maybe is worth thinking about some more.
```

Comment:
Indeed, streaming the results elsewhere may be a better design in the end, I'm not sure.  Because of the monotonicity of the entire system, we should never have to retract results, so they can be reported as soon as they are found.